### PR TITLE
IGNITE-21680 Remove destroyed RocksDB indexes on recovery

### PR DESCRIPTION
**Pull request by sashapolo**
https://issues.apache.org/jira/browse/IGNITE-21680

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java [hunk 1]
```diff
@@ -314,6 +314,113 @@ public void testDestroyIndex(boolean waitForDestroyFuture) throws Exception {
...
+                201,
+                "TEST_INDEX_2",
+                catalogTableDescriptor.id(),
+                false,
+                AVAILABLE,
+                catalogService.latestCatalogVersion(),
+                List.of(new CatalogIndexColumnDescriptor("STRKEY", ASC_NULLS_LAST))
+        );
+
+        var sortedIndex1 = new StorageSortedIndexDescriptor(catalogTableDescriptor, catalogSortedIndex1);
```
Comment (tkalkirill) [inline modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java:344]: Maybe rename to `sortedIndexDescriptor1/2` ?

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java [hunk 2]
```diff
@@ -314,6 +314,113 @@ public void testDestroyIndex(boolean waitForDestroyFuture) throws Exception {
...
+                201,
+                "TEST_INDEX_2",
+                catalogTableDescriptor.id(),
+                true,
+                AVAILABLE,
+                catalogService.latestCatalogVersion(),
+                List.of("STRKEY")
+        );
+
+        var hashIndex1 = new StorageHashIndexDescriptor(catalogTableDescriptor, catalogHashIndex1);
```
Comment (tkalkirill) [inline modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java:397]: Same about naming

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/HashIndex.java [hunk 3]
```diff
@@ -39,14 +40,23 @@ class HashIndex extends Index<RocksDbHashIndexStorage> {
...
         this.descriptor = descriptor;
         this.indexMetaStorage = indexMetaStorage;
     }
 
+    /**
+     * Returns hash index storage for partition.
+     *
+     * @param partitionId Partition ID.
+     */
+    @Nullable RocksDbHashIndexStorage getStorage(int partitionId) {
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/HashIndex.java:56]: Meow

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+                var descriptor = (StorageHashIndexDescriptor) indexDescriptorSupplier.get(indexId);
+
+                if (descriptor == null) {
+                    deleteByPrefix(writeBatch, rocksDb.hashIndexCf(), createKey(BYTE_EMPTY_ARRAY, indexId));
+                } else {
+                    hashIndices.put(indexId, new HashIndex(rocksDb, descriptor, rocksDb.meta));
+                }
+            }
+
+            var indexCfsToDestroy = new ArrayList<Map.Entry<Integer, ColumnFamily>>();
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java:76]: `<Map.Entry<Integer, ColumnFamily>>`
https://www.youtube.com/watch?v=1GiPcP30cFc

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+                    indexCfsToDestroy.add(e);
+                } else {
+                    sortedIndices.put(indexId, SortedIndex.restoreExisting(rocksDb, indexCf, descriptor, rocksDb.meta));
+                }
+            }
+
+            rocksDb.db.write(DFLT_WRITE_OPTS, writeBatch);
+
+            if (!indexCfsToDestroy.isEmpty()) {
+                rocksDb.flusher.awaitFlush(false)
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java:97]: Maybe we need to do a flush after we destroy the indexes too or not?

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+                        .thenRun(() -> {
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java:98]: Maybe runAsymc? wdyt?

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        );
+
+        return hashIndex.getOrCreateStorage(partitionStorageProvider.apply(partitionId));
+    }
+
+    @Nullable IndexStorage getIndex(int partitionId, int indexId) {
+        HashIndex hashIndex = hashIndices.get(indexId);
+
+        if (hashIndex != null) {
+            assert !sortedIndices.containsKey(indexId);
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java:133]:
```suggestion
            assert !sortedIndices.containsKey(indexId) : indexId;
```

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+            if (sortedIdx != null) {
+                sortedIdx.destroy(writeBatch);
+            }
+
+            rocksDb.db.write(DFLT_WRITE_OPTS, writeBatch);
+        }
+
+        if (sortedIdx != null) {
+            rocksDb.flusher.awaitFlush(false)
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java:193]: same about flush and runAsync

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+            hashIndex.destroy(writeBatch);
+        }
+
+        for (SortedIndex sortedIndex : sortedIndices.values()) {
+            sortedIndex.destroy(writeBatch);
+        }
+    }
+
+    void scheduleAllIndexCfDestroy() {
+        rocksDb.flusher.awaitFlush(false)
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java:219]: Same questions.

**Review by tkalkirill — APPROVED** [review]

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java (hunk 5 above)
Comment (sashapolo) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java:97]: I don't think we need to force it

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java (hunk 1 above)
Comment (sashapolo) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java]: Fixed

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java (hunk 2 above)
Comment (sashapolo) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java]: Fixed

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java (hunk 6 above)
Comment (sashapolo) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java]: Fixed

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java (hunk 7 above)
Comment (sashapolo) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java]: Fixed

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java (hunk 8 above)
Comment (sashapolo) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java:193]: Fixed

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java (hunk 9 above)
Comment (sashapolo) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java:219]: Fixed

**Review by tkalkirill — APPROVED** [review]

**Pull request merged**